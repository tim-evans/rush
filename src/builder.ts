import Constraint from "./constraint";
import SlideConstraint from "./constraints/slide";
import SnapConstraint from "./constraints/snap";

export type Orientation = "above" | "below" | "left" | "right" | "center";
export type Edge =
  | "leftEdge"
  | "rightEdge"
  | "topEdge"
  | "bottomEdge"
  | "center";

class API {
  readonly topEdge = "topEdge";
  readonly bottomEdge = "bottomEdge";
  readonly leftEdge = "leftEdge";
  readonly rightEdge = "rightEdge";
  readonly center = "center";
  orientAbove: Builder;
  orientBelow: Builder;
  orientRight: Builder;
  orientLeft: Builder;
  orientCenter: Builder;

  constructor() {
    this.orientAbove = new Builder("above");
    this.orientBelow = new Builder("below");
    this.orientRight = new Builder("right");
    this.orientLeft = new Builder("left");
    this.orientCenter = new Builder("center");
  }
}

class Builder {
  orientation: Orientation;

  constraints: Constraint[];

  next: Builder | null;

  constructor(orientation: Orientation) {
    this.orientation = orientation;
    this.constraints = [];
    this.next = null;
  }

  andSnapTo(...edges: Edge[]) {
    let orientation = this.orientation;

    this.constraints.push(
      ...edges.map(edge => {
        return new SnapConstraint({
          orientation,
          constraint: edge
        });
      })
    );

    return this;
  }

  andSlideBetween(...edges: Edge[]) {
    // Always unshift slide constraints,
    // since they should be handled first
    this.constraints.unshift(
      new SlideConstraint({
        orientation: this.orientation,
        constraints: edges
      })
    );

    return this;
  }

  where(condition: () => boolean) {
    this.constraints.forEach(constraint => {
      constraint.condition = condition;
    });

    return this;
  }

  then(next: Builder) {
    if (next !== this) {
      this.next = next;
    }

    return this;
  }
}

export default function build(
  definition: (this: API) => Builder
): Constraint[] {
  let step: Builder | null = definition.call(new API());
  let constraints = [];
  while (step) {
    constraints.push(...step.constraints);
    step = step.next;
  }

  return constraints;
}
