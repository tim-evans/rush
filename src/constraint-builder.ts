import Constraint from './constraint';
import Orientation from './orientation';

class ConstraintBuilder {
  topEdge: string;
  bottomEdge: string;
  leftEdge: string;
  rightEdge: string;
  center: string;
  orientAbove: Orientation;
  orientBelow: Orientation;
  orientRight: Orientation;
  orientLeft: Orientation;
  orientCenter: Orientation;

  constructor() {
    this.topEdge = 'top-edge';
    this.bottomEdge = 'bottom-edge';
    this.leftEdge = 'left-edge';
    this.rightEdge = 'right-edge';
    this.center = 'center';
    this.orientAbove = new Orientation('above');
    this.orientBelow = new Orientation('below');
    this.orientRight = new Orientation('right');
    this.orientLeft = new Orientation('left');
    this.orientCenter = new Orientation('center');
  }
}

interface ConstraintDefinition {
  (): Orientation
}

export default function (definition: ConstraintDefinition): Constraint[] {
  let step = definition.call(new ConstraintBuilder());
  let constraints = [];
  do {
    constraints.push(step.constraints);
    if (step.next) {
      step = step.next;
    }
  } while (step.next);

  return constraints;
}
