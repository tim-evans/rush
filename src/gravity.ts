import build from './builder';

export default {
  nw: build(function() {
    return this.orientAbove.andSnapTo(this.leftEdge);
  }),
  n: build(function() {
    return this.orientAbove.andSnapTo(this.center);
  }),
  ne: build(function() {
    return this.orientAbove.andSnapTo(this.rightEdge);
  }),
  e: build(function() {
    return this.orientRight.andSnapTo(this.center);
  }),
  se: build(function() {
    return this.orientBelow.andSnapTo(this.rightEdge);
  }),
  s: build(function() {
    return this.orientBelow.andSnapTo(this.center);
  }),
  sw: build(function() {
    return this.orientBelow.andSnapTo(this.leftEdge);
  }),
  w: build(function() {
    return this.orientLeft.andSnapTo(this.center);
  }),
  none: build(function() {
    return this.orientCenter.andSnapTo(this.center);
  })
};
