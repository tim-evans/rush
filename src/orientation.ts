import Constraint from "./constraint";

export default class Orientation {

  orientation: string;

  constraints: Constraint[];

  next: Orientation | null;

  constructor(orientation: string) {
    this.orientation = orientation;
    this.constraints = [];
    this.next = null;
  }

  andSnapTo(...guidelines) {
    let orientation = this.orientation;

    this.constraints.push(...guidelines.map((guideline) => {
      return new Constraint({ orientation, behavior: 'snap', guideline });
    }));

    return this;
  }

  andSlideBetween(...guidelines) {
    // Always unshift slide constraints,
    // since they should be handled first
    this.constraints.unshift(new Constraint({
      orientation: this.orientation,
      behavior: 'slide',
      guideline: guidelines
    }));

    return this;
  }

  where(condition) {
    this.constraints.forEach((constraint) => {
      constraint.condition = condition;
    });

    return this;
  }

  then(pipeline) {
    if (pipeline !== this) {
      this.next = guideline.constraints;
    }

    return this;
  }
}
