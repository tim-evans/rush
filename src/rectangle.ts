export default class Rectangle {
  static intersection(rectA: Rectangle, rectB: Rectangle): Rectangle {
    // Find the edges
    let x = Math.max(rectA.x, rectB.x);
    let y = Math.max(rectA.y, rectB.y);
    let right  = Math.min(rectA.right, rectB.right);
    let bottom = Math.min(rectA.bottom, rectB.bottom);
    let width = 0;
    let height = 0;

    if (rectA.right <= rectB.left ||
        rectB.right <= rectA.left ||
        rectA.bottom <= rectB.top ||
        rectB.bottom <= rectA.top) {
      x = y = width = height = 0;
    } else {
      width  = Math.max(0, right - x);
      height = Math.max(0, bottom - y);
    }

    return new Rectangle(x, y, width, height);
  }

  static fromDOMRect(rect: DOMRect | ClientRect): Rectangle {
    if (rect instanceof DOMRect) {
      return new Rectangle(rect.x, rect.y, rect.width, rect.height);
    }
    return new Rectangle(rect.left, rect.top, rect.width, rect.height);
  }

  left: number;
  top: number;
  width: number;
  height: number;

  constructor(x: number, y: number, width: number, height: number) {
    this.left = x;
    this.top = y;
    this.width = width;
    this.height = height;
  }

  get right(): number {
    return this.left + this.width;
  }

  get bottom(): number {
    return this.top + this.height;
  }

  get area(): number {
    return this.width * this.height;
  }

  intersects(rect: Rectangle): boolean {
    return Rectangle.intersection(this, rect).area > 0;
  }

  contains(rect: Rectangle): boolean {
    return Rectangle.intersection(this, rect).area === rect.area;
  }

  get center(): [number, number] {
    return [this.left + this.width / 2, this.top + this.height / 2];
  }

  translateX(dX: number) {
    this.left = this.left + dX;
  }

  translateY(dY: number) {
    this.top = this.top + dY;
  }

  translate(dX: number, dY: number) {
    this.translateX(dX);
    this.translateY(dY);
  }

  get x(): number {
    return this.left;
  }

  set x(value: number) {
    this.translateX(value - this.left);
  }

  get y(): number {
    return this.top;
  }

  set y(value: number) {
    this.translateY(value - this.top);
  }
}
