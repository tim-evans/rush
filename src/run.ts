import Rectangle from './rectangle';
import Constraint from './constraint';
import { getLayout } from 'dom-ruler';

function run(constraints: Constraint[], targetArea: DOMRect, $popover: HTMLElement, $pointer?: HTMLElement) {
  let $container = scrollParent($popover);
  let target = Rectangle.fromDOMRect(targetArea);
  let popover = new Rectangle(0, 0, 0, 0);
  let pointer = new Rectangle(0, 0, 0, 0);

  let solution = constraints.find(function (constraint) {
    constraint.solveFor(boundary, target, popover);
  });

  let isHidden = Rectangle.intersection(boundingRect, targetRect).area === 0;

  if (boundingElement === document.scrollingElement) {
    popover.translateY(-1, $container.scrollTop);
    popover.translateX(-1, $container.scrollLeft);
    target.translateY(-1, $container.scrollTop);
    target.translateX(-1, $container.scrollLeft);
  }

  if (animated) {
    let [centerX, centerY] = target.center();
    $container.style.top = centerY + 'px';
    $container.style.left = centerX + 'px';

    $popover.style.top = (popover.top - centerY) + 'px';
    $popover.style.left = (popover.left - centerX) + 'px';
    $popover.style.width = popover.width + 'px';
    $popover.style.height = popover.height + 'px';
  } else {
    $popover.style.top = popover.top + 'px';
    $popover.style.left = popover.left + 'px';
    $popover.style.width = popover.width + 'px';
    $popover.style.height = popover.height + 'px';
  }

  if ($pointer) {
    $pointer.style.top = pointer.top + 'px';
    $pointer.style.left = pointer.left + 'px';
  }
}

function scrollParent(element: HTMLElement): HTMLElement {
  let position = element.style.position;
  let excludeStaticParent = position === 'absolute';
  let parent = element.parentElement;

  while (parent) {
    if (excludeStaticParent && parent.style.position === 'static') {
      parent = parent.parentElement;
      continue;
    }

    let { overflow, overflowX, overflowY } = parent.style;
    if (/(auto|scroll)/.test(overflow + overflowX + overflowY)) {
      break;
    }
    parent = parent.parentElement;
  }

  if (parent == null) {
    parent = document.scrollingElement;
  }

  return position === 'fixed' || parent;
}
