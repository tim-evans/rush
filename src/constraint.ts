import { Edge, Orientation } from './builder';
import Rectangle from './rectangle';

export default abstract class Constraint {
  condition?: (target: Rectangle, popover: Rectangle, pointer: Rectangle) => boolean;

  abstract solveFor(
    boundingRect: Rectangle,
    targetRect: Rectangle,
    popoverRect: Rectangle,
    pointerRect: Rectangle,
    positionOver: boolean
  ): {
    orientation: Orientation;
    edge: Edge;
    valid: boolean;
  };

  horizontallyCenter(target: Rectangle, popover: Rectangle, pointer: Rectangle) {
    popover.x = target.left + target.width / 2 - popover.width / 2;
    pointer.x = popover.width / 2 - pointer.width / 2;
  }

  verticallyCenter(target: Rectangle, popover: Rectangle, pointer: Rectangle) {
    popover.y = target.top + target.height / 2 - popover.height / 2;
    pointer.y = popover.height / 2 - pointer.height / 2;
  }

  positionAbove(target: Rectangle, popover: Rectangle, pointer: Rectangle, positionOver: boolean) {
    let shiftY = positionOver ? 0 : popover.height + pointer.height;
    popover.y = target.top - shiftY;
    pointer.y = popover.height;
  }

  positionBelow(target: Rectangle, popover: Rectangle, pointer: Rectangle, positionOver: boolean) {
    let shiftY = positionOver ? -1 * popover.height :  pointer.height;
    popover.y = target.bottom + shiftY;
    pointer.y = pointer.height * -1;
  }

  positionLeft(target: Rectangle, popover: Rectangle, pointer: Rectangle, positionOver: boolean) {
    let shiftX = positionOver ? 0 : popover.width + pointer.width;
    popover.x = target.left - shiftX;
    pointer.x = popover.width;
  }

  positionRight(target: Rectangle, popover: Rectangle, pointer: Rectangle, positionOver: boolean) {
    const shiftX = positionOver ? (-1 * popover.width) : pointer.width;
    popover.x = target.right + shiftX;
    pointer.x = pointer.width * -1;
  }

  positionCenter(target: Rectangle, popover: Rectangle, pointer: Rectangle) {
    this.horizontallyCenter(target, popover, pointer);
    this.verticallyCenter(target, popover, pointer);
  }
}
