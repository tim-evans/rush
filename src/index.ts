import build from "./builder";
import Constraint from "./constraint";
import SlideConstraint from "./constraints/slide";
import SnapConstraint from "./constraints/snap";
import gravity from "./gravity";
import Rectangle from "./rectangle";

export function scrollParent(element: HTMLElement): Element | HTMLElement {
  let position = element.style.position;
  let excludeStaticParent = position === "absolute";
  let parent = element.parentElement;

  while (parent) {
    if (excludeStaticParent && parent.style.position === "static") {
      parent = parent.parentElement;
      continue;
    }

    let { overflow, overflowX, overflowY } = window.getComputedStyle(parent);
    if (/(auto|scroll)/.test((overflow || "") + overflowX + overflowY)) {
      break;
    }
    parent = parent.parentElement;
  }

  if (parent == null) {
    return document.scrollingElement || document.documentElement;
  }

  return parent;
}

export const solveFor = (
  constraints: Constraint[],
  boundingRect: DOMRect | ClientRect | Rectangle,
  targetRect: DOMRect | ClientRect | Rectangle,
  popoverRect: DOMRect | ClientRect | Rectangle,
  pointerRect?: DOMRect | ClientRect | Rectangle
) => {
  let target = Rectangle.fromDOMRect(targetRect);
  let bounding = Rectangle.fromDOMRect(boundingRect);
  let popover = Rectangle.fromDOMRect(popoverRect);
  let pointer = new Rectangle(0, 0, 0, 0);
  if (pointerRect) {
    pointer = Rectangle.fromDOMRect(pointerRect);
  }

  let constraint =
    constraints.find(option => {
      return option.solveFor(bounding, target, popover, pointer, false).valid;
    }) || constraints[constraints.length - 1];
  let solution = constraint.solveFor(bounding, target, popover, pointer, false);

  return {
    ...solution,
    popoverRect: popover,
    pointerRect: pointer
  };
};

export {
  build,
  gravity,
  Rectangle,
  Constraint,
  SlideConstraint,
  SnapConstraint
};
