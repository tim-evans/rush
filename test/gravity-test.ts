import * as gravity from '../src/gravity';

describe('gravity', () => {
  test.each(
    Object.keys(gravity)
  )('%s', direction => {
    expect((gravity as any)[direction]).toMatchSnapshot();
  });
});
