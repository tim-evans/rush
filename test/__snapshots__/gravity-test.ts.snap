// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gravity default 1`] = `
Object {
  "e": Array [
    SnapConstraint {
      "constraint": "center",
      "orientation": "right",
    },
  ],
  "n": Array [
    SnapConstraint {
      "constraint": "center",
      "orientation": "above",
    },
  ],
  "ne": Array [
    SnapConstraint {
      "constraint": "rightEdge",
      "orientation": "above",
    },
  ],
  "none": Array [
    SnapConstraint {
      "constraint": "center",
      "orientation": "center",
    },
  ],
  "nw": Array [
    SnapConstraint {
      "constraint": "leftEdge",
      "orientation": "above",
    },
  ],
  "s": Array [
    SnapConstraint {
      "constraint": "center",
      "orientation": "below",
    },
  ],
  "se": Array [
    SnapConstraint {
      "constraint": "rightEdge",
      "orientation": "below",
    },
  ],
  "sw": Array [
    SnapConstraint {
      "constraint": "leftEdge",
      "orientation": "below",
    },
  ],
  "w": Array [
    SnapConstraint {
      "constraint": "center",
      "orientation": "left",
    },
  ],
}
`;
